### ISSUE TITLE
Issues installing nginx-ingress on AKS cluster

### ISSUE THREAD

**Issue by shashank-g172** [microbean/microbean-helm#187]
I created a brand new AKS cluster, and configured the KubernetesClient with the config object using the access token and the API. The cluster has RBAC disabled. 

When I try to follow the example on how to install the chart as given in: https://microbean.github.io/microbean-helm/examples.html (using wordpress as the example chart)

I see the following error:
```
Could not forward port to a Ready Tiller pod's port 44134 in namespace kube-system with labels {app=helm, name=tiller}
```

Next, I manually installed the tiller in the AKS cluster by following these steps: https://docs.microsoft.com/en-us/azure/aks/kubernetes-helm (created the serviceAccount, helm init, and helm search). They all worked. 

When I try to install the nginx-ingress chart using the microbean Helm API, I see this error:
```
Caused by: java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: UNKNOWN: release random-nginx failed: clusterroles.rbac.authorization.k8s.io "random-nginx-nginx-ingress" is forbidden: attempt to grant extra privileges: [{[list] [] [configmaps] [] []} {[watch] [] [configmaps] [] []} {[list] [] [endpoints] [] []} {[watch] [] [endpoints] [] []} {[list] [] [nodes] [] []} {[watch] [] [nodes] [] []} {[list] [] [pods] [] []} {[watch] [] [pods] [] []} {[list] [] [secrets] [] []} {[watch] [] [secrets] [] []} {[get] [] [nodes] [] []} {[get] [] [services] [] []} {[list] [] [services] [] []} {[update] [] [services] [] []} {[watch] [] [services] [] []} {[get] [extensions] [ingresses] [] []} {[list] [extensions] [ingresses] [] []} {[watch] [extensions] [ingresses] [] []} {[create] [] [events] [] []} {[patch] [] [events] [] []} {[update] [extensions] [ingresses/status] [] []}] user=&{system:serviceaccount:kube-system:default e91ea5b0-e128-11e8-8d0d-524a19eede43 [system:serviceaccounts system:serviceaccounts:kube-system system:authenticated] map[]} ownerrules=[] ruleResolutionErrors=[]
	at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:500)
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:479)
	at com.mathworks.test.tools.fixtures.ephemeral.deployer.helm.HelmUtils.pullNginxController(HelmUtils.java:57)
	at com.mathworks.test.tools.fixtures.ephemeral.deployer.helm.HelmUtils.main(HelmUtils.java:73)
	... 6 more
Caused by: io.grpc.StatusRuntimeException: UNKNOWN: release random-nginx failed: clusterroles.rbac.authorization.k8s.io "random-nginx-nginx-ingress" is forbidden: attempt to grant extra privileges: [{[list] [] [configmaps] [] []} {[watch] [] [configmaps] [] []} {[list] [] [endpoints] [] []} {[watch] [] [endpoints] [] []} {[list] [] [nodes] [] []} {[watch] [] [nodes] [] []} {[list] [] [pods] [] []} {[watch] [] [pods] [] []} {[list] [] [secrets] [] []} {[watch] [] [secrets] [] []} {[get] [] [nodes] [] []} {[get] [] [services] [] []} {[list] [] [services] [] []} {[update] [] [services] [] []} {[watch] [] [services] [] []} {[get] [extensions] [ingresses] [] []} {[list] [extensions] [ingresses] [] []} {[watch] [extensions] [ingresses] [] []} {[create] [] [events] [] []} {[patch] [] [events] [] []} {[update] [extensions] [ingresses/status] [] []}] user=&{system:serviceaccount:kube-system:default e91ea5b0-e128-11e8-8d0d-524a19eede43 [system:serviceaccounts system:serviceaccounts:kube-system system:authenticated] map[]} ownerrules=[] ruleResolutionErrors=[]
	at io.grpc.Status.asRuntimeException(Status.java:526)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:467)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:41)
	at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:684)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:41)
	at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:391)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:475)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:557)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:478)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:590)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
```
This is the code I'm executing:
```
public static DefaultKubernetesClient pullNginxController(Config config) throws Exception {
    // String nginxChart = "https://kubernetes-charts.storage.googleapis.com/nginx-ingress-0.30.0.tgz";
    String wordPressChart = "https://kubernetes-charts.storage.googleapis.com/wordpress-0.6.6.tgz";
   URI uri = URI.create(wordPressChart);
    URL url = uri.toURL();

    Chart.Builder chart = null;
    try (final URLChartLoader chartLoader = new URLChartLoader()) {
      chart = chartLoader.load(url);
    }
    assert chart != null;

    try (final DefaultKubernetesClient client = new DefaultKubernetesClient(config);
        Tiller tiller = new Tiller(client);
        final ReleaseManager releaseManager = new ReleaseManager(tiller)) {

      final InstallReleaseRequest.Builder requestBuilder = InstallReleaseRequest.newBuilder();
      assert requestBuilder != null;
      requestBuilder.setTimeout(800L);
      requestBuilder.setName("random-nginx"); // Set the Helm release name
      requestBuilder.setWait(true); // Wait for Pods to be ready

      // Install the loaded chart with the user-supplied overrides.
      final Future<InstallReleaseResponse> releaseFuture =
          releaseManager.install(requestBuilder, chart);
      assert releaseFuture != null;
      final Release release = releaseFuture.get().getRelease();
      assert release != null;
      
      return client;
    }
  }
```
I can now deploy the wordpress chart to the default namespace using this code. 

My two questions are:

1. How do I get past the port forwarding error without having to manually install helm and the tiller pod in the AKS cluster? I'm on the 2.8.2.1.1.1 version of microbean-helm. 
2. This may not be directly related to the Helm API project, but as far as permissions go - there are more resources being created in the nginx chart than the wordpress chart - is there an example of Helm Init using this API available?

Thanks,
Shashank

**Comment by ljnelson** [comment]
Hi; thanks for using microBean Helm.

> My two questions are:
> 
> 1. How do I get past the port forwarding error without having to manually install helm and the tiller pod in the AKS cluster? I'm on the 2.8.2.1.1.1 version of microbean-helm.
> 2. This may not be directly related to the Helm API project, but as far as permissions go - there are more resources being created in the nginx chart than the wordpress chart - is there an example of Helm Init using this API available?

Regarding your first question:

microBean Helm models the client part of the client-server stew that the `helm` command line program comprises.  So if there is no server&mdash;i.e. no Tiller&mdash; for a Helm client (`helm` or microBean Helm) to talk to, then there won't be any Pod to forward commands to.

Your question therefore reduces to: "How can I install Tiller in the cluster prior to using microBean Helm?"

You can do this a couple of different ways:

1. Since installation of Tiller is a one-time operation, you may find it easiest to simply use the `helm` command line tool and then forget about it forevermore.  See [`helm help init`](https://docs.helm.sh/helm/#helm-init).
2. Alternatively if for whatever reason you'd like to also perform this action using microBean Helm, you can install Tiller using the [`TillerInstaller`](https://microbean.github.io/microbean-helm/apidocs/org/microbean/helm/TillerInstaller.html) class.  Please note as the documentation says that this class is experimental and the API is subject to change without notice.

With regard to your second question, I'm not really sure what you're asking.  At chart installation time, initialization is a thing of the past, so there is no example using microBean Helm of doing any kind of initialization behavior&mdash;i.e. installing Tiller&mdash;as part of a chart installation, nginx or otherwise.

If you are asking a general question about Helm _charts_, whose structure and definition are mostly outside the scope of microBean Helm, you might have a look at the [chart documentation](https://docs.helm.sh/developing_charts/#charts).

**Comment by shashank-g172** [comment]
Thank you @ljnelson! That explains it more. An enhancement request may come out of this, the microbean-helm right now uses an older version of the kubernetes-client (3.1.8) but the latest one is 4.1.0. In order to use the latest features of the kuberentes client library I've managed my dependencies to use the 4.1.0 of the library. 

When the TillerInstaller looks for the Deployment class, it's looking in this path:
import io.fabric8.kubernetes.api.model.extensions.Deployment;

But the latest version of the library (4.1.0) has it here:
import io.fabric8.kubernetes.api.model.apps.Deployment. 

Would it be possible to update the project with the latest version of the kubernetes-client JAR?

As for the second question, yeah I figured it was a little out of the Helm scope - I think I faced that issue because of RAB being disabled.
